service/storage_service.cc

```diff
@@ -721,27 +721,55 @@ future<> storage_service::topology_transition() {
 }
 
 future<> storage_service::merge_topology_snapshot(raft_topology_snapshot snp) {
-    std::vector<mutation> muts;
-    muts.reserve(snp.topology_mutations.size() + snp.cdc_generation_mutations.size() + snp.topology_requests_mutations.size());
-    {
-        auto s = _db.local().find_schema(db::system_keyspace::NAME, db::system_keyspace::TOPOLOGY);
-        boost::transform(snp.topology_mutations, std::back_inserter(muts), [s] (const canonical_mutation& m) {
-            return m.to_mutation(s);
-        });
-    }
-    if (snp.cdc_generation_mutations.size() > 0) {
+    if (!snp.cdc_generation_mutations.empty()) {
         auto s = _db.local().find_schema(db::system_keyspace::NAME, db::system_keyspace::CDC_GENERATIONS_V3);
-        boost::transform(snp.cdc_generation_mutations, std::back_inserter(muts), [s] (const canonical_mutation& m) {
-            return m.to_mutation(s);
+
+        // Split big mutations into smaller ones, prepare frozen_muts_to_apply
+        std::vector<frozen_mutation> frozen_muts_to_apply;
+        {
+            std::vector<mutation> muts_to_apply;
+            muts_to_apply.reserve(snp.cdc_generation_mutations.size());
+            const auto max_size = _db.local().commitlog()->max_record_size() / 2;
+            for (const auto& m: snp.cdc_generation_mutations) {
+                auto mut = m.to_mutation(s);
+                if (m.representation().size() <= max_size) {
+                    muts_to_apply.push_back(std::move(mut));
+                } else {
+                    co_await split_mutation(std::move(mut), muts_to_apply, max_size);
+                }
+            }
+            frozen_muts_to_apply = freeze(muts_to_apply);
+        }
+
+        // Apply non-atomically so as not to hit the commitlog size limit.
+        // The cdc_generations_v3 data is not used in any way until
+        // it's referenced from the topology table.
+        // By applying the cdc_generations_v3 mutations before topology mutations
+        // we ensure that the lack of atomicity isn't a problem here.
+        co_await max_concurrent_for_each(frozen_muts_to_apply, 128, [&] (const frozen_mutation& m) -> future<> {
+            return _db.local().apply(s, m, {}, db::commitlog::force_sync::no, db::no_timeout);
```

Comment:
Although it probably doesn't matter what `force_sync` you pass here (since the table uses schema commitlog), but using `force_sync::no` is confusing. (And we do need to sync.)

Reply:
done